flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SsgNetworkMemoryCalculationUtils.java

```diff
@@ -148,6 +167,80 @@ private static TaskInputsOutputsDescriptor buildTaskInputsOutputsDescriptor(
         return ret;
     }
 
+    private static Map<IntermediateDataSetID, Integer> getMaxInputChannelNumsForDynamicGraph(
+            ExecutionJobVertex ejv) {
+
+        Map<IntermediateDataSetID, Integer> ret = new HashMap<>();
+
+        final Map<Tuple2<IntermediateDataSetID, ExecutionVertexID>, ConsumedPartitionGroup>
+                consumedPartitionGroups = new HashMap<>();
+        for (ExecutionVertex vertex : ejv.getTaskVertices()) {
+            for (ConsumedPartitionGroup partitionGroup : vertex.getAllConsumedPartitionGroups()) {
+                consumedPartitionGroups.put(
+                        new Tuple2<>(partitionGroup.getIntermediateDataSetID(), vertex.getID()),
+                        partitionGroup);
+            }
+        }
+
+        for (IntermediateResult consumedResult : ejv.getInputs()) {
+            ret.put(
+                    consumedResult.getId(),
+                    getMaxInputChannelNumForResult(
+                            ejv,
+                            consumedResult.getId(),
+                            (resultId, vertexId) ->
+                                    consumedPartitionGroups.get(new Tuple2<>(resultId, vertexId))));
+        }
+
+        return ret;
+    }
+
+    private static Map<IntermediateDataSetID, Integer> getMaxSubpartitionNumsForDynamicGraph(
+            ExecutionJobVertex ejv) {
+
+        Map<IntermediateDataSetID, Integer> ret = new HashMap<>();
+
+        for (IntermediateResult intermediateResult : ejv.getProducedDataSets()) {
+            final int maxNum =
+                    Arrays.stream(intermediateResult.getPartitions())
+                            .map(IntermediateResultPartition::getNumberOfSubpartitions)
+                            .reduce(0, Integer::max);
+            ret.put(intermediateResult.getId(), maxNum);
+        }
+
+        return ret;
+    }
+
+    @VisibleForTesting
+    static int getMaxInputChannelNumForResult(
```

Comment:
I would change the method to `Map< IntermediateDataSetID, Integer> getMaxInputChannelNumForJobVertex(ExecutionJobVertex)`. In this way, we do not need to pre-build the map for retriever and `testGetMaxInputChannelNumForResult()` can be simplified.

Reply:
Do you mean to merge `getMaxInputChannelNumForResult ` into `getMaxSubpartitionNumsForDynamicGraph` ?

Reply:
You are right.